README.md
<h1 align="center">
  <br>
    <img src="assets/icon.png" alt="logo" width="150">
  <br><br>
  Indent Colorizer
  <br>
  <br>
</h1>
<p align="center">
    <a href="https://marketplace.visualstudio.com/items?itemName=tal7aouy.indent-colorizer"><img src="https://vsmarketplacebadge.apphb.com/version-short/tal7aouy.indent-colorizer.svg?style=for-the-badge&colorA=252526&colorB=43A047&label=VERSION" alt="Version"></a>&nbsp;
    <a href="https://marketplace.visualstudio.com/items?itemName=tal7aouy.indent-colorizer"><img src="https://vsmarketplacebadge.apphb.com/rating-short/tal7aouy.indent-colorizer.svg?style=for-the-badge&colorA=252526&colorB=43A047&label=Rating" alt="Rating"></a>&nbsp;
    <a href="https://marketplace.visualstudio.com/items?itemName=tal7aouy.indent-colorizer"><img src="https://vsmarketplacebadge.apphb.com/installs-short/tal7aouy.indent-colorizer.svg?style=for-the-badge&
    colorA=252526&colorB=43A047&label=Installs" alt="Installs"></a>&nbsp;
    <a href="https://marketplace.visualstudio.com/items?itemName=tal7aouy.indent-colorizer"><img src="https://vsmarketplacebadge.apphb.com/downloads-short/tal7aouy.indent-colorizer.svg?style=for-the-badge&colorA=252526&colorB=43A047&label=Downloads" alt="Downloads"></a>
</p>

---

## Installation

All instructions can be found at [INSTALL.md](./INSTALL.md).

A simple extension to make indentation more readable

![ScreenShot](https://raw.githubusercontent.com/tal7aouy/vscode-indent-colorizer/master/assets/screen.png)

Get it here: [Visual Studio Code Marketplace](https://marketplace.visualstudio.com/items?itemName=tal7aouy.indent-colorizer)

It uses the current editor window tab size and can handle mixed tab + spaces but that is not recommended. In addition it visibly marks lines where the indentation is not a multiple of the tab size. This should help to find problems with indentation in some situations.

### Configuration

Although you can just use it as it is there is the possibility to configure some aspects of the extension:

```js
  // For which languages indent-colorizer should be activated (if empty it means all).
  "indentColorizer.includedLanguages": [] // for example ["php", "ts", "python"]

  // For which languages indent-colorizer should be deactivated (if empty it means none).
  "indentColorizer.excludedLanguages": ["plaintext"]

  // The delay in ms until the editor gets updated.
  "indentColorizer.updateDelay": 100 // 10 makes it super fast but may cost more resources
```

_Notice: Defining both `includedLanguages` and `excludedLanguages` does not make much sense. Use one of both!_

You can configure your own colors by adding and tampering with the following code:

```js
  // Defining custom colors instead of default "Colorizer" for dark backgrounds.
  // (Sorry: Changing them needs an editor restart for now!)
  "indentColorizer.colors": [
    "rgba(255,255,64,0.07)",
    "rgba(127,255,205,0.07)",
    "rgba(255,128,210,0.07)",
    "rgba(79,236,236,0.07)",
    "rgba(78,184,237,0.07)",
    "rgba(233,226,74,0.07)",
    "rgba(101,74,233,0.07)",
  ]

  // The indent color if the number of spaces is not a multiple of "tabSize".
  "indentColorizer.errorColor": "rgba(128,32,32,0.6)"

  // The indent color when there is a mix between spaces and tabs.
  // To be disabled this coloring set this to an empty string.
  "indentColorizer.tabmixColor": "rgba(128,32,96,0.6)"
```

Skip error highlighting for RegEx patterns. For example, you may want to turn off the indent errors for JSDoc's valid additional space (disabled by default), or comment lines beginning with `//`

```js
  // Example of regular expression in JSON (note double backslash to escape characters)
  "indentColorizer.ignoreLinePatterns" : [
    "/[ \t]* [*]/g", // lines begining with <whitespace><space>*
    "/[ \t]+[/]{2}/g" // lines begininning with <whitespace>//
  ]
```

Skip error highlighting for some or all languages. For example, you may want to turn off the indent errors for `markdown` and `haskell` (which is the default)

```js
  "indentColorizer.ignoreErrorLanguages" : [
    "markdown",
    "haskell"
  ]
```

If error color is disabled, indent colors will be rendered until the length of rendered characters (white spaces, tabs, and other ones) is divisible by tabsize. Turn on this option to render white spaces and tabs only.

```js
  "indentColorizer.colorOnWhiteSpaceOnly": true // false is the default
```


CHANGELOG
# v1.0.0

- initial release


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";

// this method is called when vs code is activated
export function activate(context: vscode.ExtensionContext) {
 // Create a decorator types that we use to decorate indent levels
 let decorationTypes = [];

 let doIt = false;
 let clearMe = false;
 let currentLanguageId = null;
 let skipAllErrors = false;

 let activeEditor = vscode.window.activeTextEditor;

 // Error color gets shown when tabs aren't right,
 //  e.g. when you have your tabs set to 2 spaces but the indent is 3 spaces
 const error_color =
  vscode.workspace.getConfiguration("indentColorizer")["errorColor"] ||
  "rgba(128,32,32,0.3)";
 const error_decoration_type = vscode.window.createTextEditorDecorationType({
  backgroundColor: error_color,
 });

 const tabmix_color =
  vscode.workspace.getConfiguration("indentColorizer")["tabmixColor"] || "";
 const tabmix_decoration_type =
  "" !== tabmix_color
   ? vscode.window.createTextEditorDecorationType({
      backgroundColor: tabmix_color,
     })
   : null;

 const ignoreLinePatterns =
  vscode.workspace.getConfiguration("indentColorizer")["ignoreLinePatterns"] ||
  [];
 const colorOnWhiteSpaceOnly =
  vscode.workspace.getConfiguration("indentColorizer")[
   "colorOnWhiteSpaceOnly"
  ] || false;

 // Colors will cycle through, and can be any size that you want
 const colors = vscode.workspace.getConfiguration("indentColorizer")[
  "colors"
 ] || [
  "rgba(255,255,64,0.07)",
  "rgba(127,255,205,0.07)",
  "rgba(255,128,210,0.07)",
  "rgba(79,236,236,0.07)",
  "rgba(78,184,237,0.07)",
  "rgba(233,226,74,0.07)",
  "rgba(101,74,233,0.07)",
 ];

 // Loops through colors and creates decoration types for each one
 colors.forEach((color, index) => {
  decorationTypes[index] = vscode.window.createTextEditorDecorationType({
   backgroundColor: color,
  });
 });

 // loop through ignore regex strings and convert to valid RegEx's.
 ignoreLinePatterns.forEach((ignorePattern, index) => {
  if (typeof ignorePattern === "string") {
   //parse the string for a regex
   var regParts = ignorePattern.match(/^\/(.*?)\/([gim]*)$/);
   if (regParts) {
    // the parsed pattern had delimiters and modifiers. handle them.
    ignoreLinePatterns[index] = new RegExp(regParts[1], regParts[2]);
   } else {
    // we got pattern string without delimiters
    ignoreLinePatterns[index] = new RegExp(ignorePattern);
   }
  }
 });

 if (activeEditor) {
  indentConfig();
 }

 if (activeEditor && checkLanguage()) {
  triggerUpdateDecorations();
 }

 vscode.window.onDidChangeActiveTextEditor(
  (editor) => {
   activeEditor = editor;
   if (editor) {
    indentConfig();
   }

   if (editor && checkLanguage()) {
    triggerUpdateDecorations();
   }
  },
  null,
  context.subscriptions
 );

 vscode.workspace.onDidChangeTextDocument(
  (event) => {
   if (activeEditor) {
    indentConfig();
   }

   if (
    activeEditor &&
    event.document === activeEditor.document &&
    checkLanguage()
   ) {
    triggerUpdateDecorations();
   }
  },
  null,
  context.subscriptions
 );

 function isEmptyObject(obj) {
  return Object.getOwnPropertyNames(obj).length === 0;
 }

 function indentConfig() {
  var skiplang =
   vscode.workspace.getConfiguration("indentRainbow")["ignoreErrorLanguages"] ||
   [];
  skipAllErrors = false;
  if (skiplang.length !== 0) {
   if (
    skiplang.indexOf("*") !== -1 ||
    skiplang.indexOf(currentLanguageId) !== -1
   ) {
    skipAllErrors = true;
   }
  }
 }

 function checkLanguage() {
  if (activeEditor) {
   if (currentLanguageId !== activeEditor.document.languageId) {
    var inclang =
     vscode.workspace.getConfiguration("indentRainbow")["includedLanguages"] ||
     [];
    var exclang =
     vscode.workspace.getConfiguration("indentRainbow")["excludedLanguages"] ||
     [];

    currentLanguageId = activeEditor.document.languageId;
    doIt = true;
    if (inclang.length !== 0) {
     if (inclang.indexOf(currentLanguageId) === -1) {
      doIt = false;
     }
    }

    if (doIt && exclang.length !== 0) {
     if (exclang.indexOf(currentLanguageId) !== -1) {
      doIt = false;
     }
    }
   }
  }

  if (clearMe && !doIt) {
   // Clear decorations when language switches away
   var decor: vscode.DecorationOptions[] = [];
   for (let decorationType of decorationTypes) {
    activeEditor.setDecorations(decorationType, decor);
   }
   clearMe = false;
  }

  indentConfig();

  return doIt;
 }

 var timeout = null;
 function triggerUpdateDecorations() {
  if (timeout) {
   clearTimeout(timeout);
  }
  var updateDelay =
   vscode.workspace.getConfiguration("indentRainbow")["updateDelay"] || 100;
  timeout = setTimeout(updateDecorations, updateDelay);
 }

 function updateDecorations() {
  if (!activeEditor) {
   return;
  }
  var regEx = /^[\t ]+/gm;
  var text = activeEditor.document.getText();
  var tabSizeRaw = activeEditor.options.tabSize;
  var tabSize = 4;
  if (tabSizeRaw !== "auto") {
   tabSize = +tabSizeRaw;
  }
  var tabs = " ".repeat(tabSize);
  const ignoreLines = [];
  let error_decorator: vscode.DecorationOptions[] = [];
  let tabmix_decorator: vscode.DecorationOptions[] = tabmix_decoration_type
   ? []
   : null;
  let decorators = [];
  decorationTypes.forEach(() => {
   let decorator: vscode.DecorationOptions[] = [];
   decorators.push(decorator);
  });

  var match;
  var ignore;

  if (!skipAllErrors) {
   /**
    * Checks text against ignore regex patterns from config(or default).
    * stores the line positions of those lines in the ignoreLines array.
    */
   ignoreLinePatterns.forEach((ignorePattern) => {
    while ((ignore = ignorePattern.exec(text))) {
     const pos = activeEditor.document.positionAt(ignore.index);
     const line = activeEditor.document.lineAt(pos).lineNumber;
     ignoreLines.push(line);
    }
   });
  }

  var re = new RegExp("\t", "g");
  let defaultIndentCharRegExp = null;

  while ((match = regEx.exec(text))) {
   const pos = activeEditor.document.positionAt(match.index);
   const line = activeEditor.document.lineAt(pos).lineNumber;
   let skip = skipAllErrors || ignoreLines.indexOf(line) !== -1; // true if the lineNumber is in ignoreLines.
   var thematch = match[0];
   var ma = match[0].replace(re, tabs).length;
   /**
    * Error handling.
    * When the indent spacing (as spaces) is not divisible by the tabsize,
    * consider the indent incorrect and mark it with the error decorator.
    * Checks for lines being ignored in ignoreLines array ( `skip` Boolran)
    * before considering the line an error.
    */
   if (!skip && ma % tabSize !== 0) {
    var startPos = activeEditor.document.positionAt(match.index);
    var endPos = activeEditor.document.positionAt(
     match.index + match[0].length
    );
    var decoration = {
     range: new vscode.Range(startPos, endPos),
     hoverMessage: null,
    };
    error_decorator.push(decoration);
   } else {
    var m = match[0];
    var l = m.length;
    var o = 0;
    var n = 0;
    while (n < l) {
     const s = n;
     var startPos = activeEditor.document.positionAt(match.index + n);
     if (m[n] === "\t") {
      n++;
     } else {
      n += tabSize;
     }
     if (colorOnWhiteSpaceOnly && n > l) {
      n = l;
     }
     var endPos = activeEditor.document.positionAt(match.index + n);
     var decoration = {
      range: new vscode.Range(startPos, endPos),
      hoverMessage: null,
     };
     var sc = 0;
     var tc = 0;
     if (!skip && tabmix_decorator) {
      // counting (split is said to be faster than match()
      // only do it if we don't already skip all errors
      var tc = thematch.split("\t").length - 1;
      if (tc) {
       // only do this if we already have some tabs
       var sc = thematch.split(" ").length - 1;
      }
      // if we have (only) "spaces" in a "tab" indent file we
      // just ignore that, because we don't know if there
      // should really be tabs or spaces for indentation
      // If you (yes you!) know how to find this out without
      // infering this from the file, speak up :)
     }
     if (sc > 0 && tc > 0) {
      tabmix_decorator.push(decoration);
     } else {
      let decorator_index = o % decorators.length;
      decorators[decorator_index].push(decoration);
     }
     o++;
    }
   }
  }
  decorationTypes.forEach((decorationType, index) => {
   activeEditor.setDecorations(decorationType, decorators[index]);
  });
  activeEditor.setDecorations(error_decoration_type, error_decorator);
  tabmix_decoration_type &&
   activeEditor.setDecorations(tabmix_decoration_type, tabmix_decorator);
  clearMe = true;
 }
}
